content/en/contribute/code/core/automated-tests.md

```diff
@@ -117,3 +126,24 @@ Running e2e tests can be quite slow so to save time modify the `specs` property
+
+## Balancing effort and outcomes
+We seek to have a quality codebase that developers can work on with speed. This means balancing test strategies, quantity, and coverage.
+
+When looking at a well-factored codebase there there are three common ways to automate tests:
+1. Unit tests
+1. Integration tests
+1. end-to-end tests (sometimes called "feature" or "system" tests)
+
+The different test types tend to increase in cost, complexity, time to run, and fragility in the order listed above. Below is a table outlining these factors:
+
+
+| Test Type         | Description| Expectations | Execution Speed     | Complexity | Fragility | 
+|--------------|-----------------|---------------------|------------|---------|---------| 
+| Unit  | Small tests of specific behavior. Often written to handle a specific condition within a specific function/component. Any dependencies are often mocked.  | High coverage of functionality. If measured in branch coverage percentage, aim for >= 90%. | Extremely fast | Extremely low | Extremely stable |
+| Integration  | Tests to exercise how multiple components interact with each other. With a dynamic language like JavaScript these are especially important to verify expectations of interface points. These may mock some parts, but often use the "real" components since the point is to exercise those components together. As a result, these tests likely involve more setup, potentially involving data scenarios. | Dramatically fewer than unit tests. The goal is not to verify all branches; it is to gain confidence in interface points.  | Fast execution, but slower startup when working with a DB | Mid-to-high. Things can get complex fast when combining parts! | Mostly stable. Fragility risks tend to come from DB setup. |
+| End-to-end  | Verify that major system features work as expected. This means things like logging in, submitting a form, showing/hiding a tab based on configuration. Another way of thinking of it is that it's a "sanity check". | Very few. This isn't to test every single thing or even close to it. Something close to 20 to 30 (total!) is reasonable. This is not the place to guarantee confidence in the system. The two types above are for that. | Painfully slow. So slow that too many of them can lead to developers not bothering to run them! | Low for the test itself (click tab, enter text into form, click submit, check text on screen. Extremely high for the setup. | Painful fragility with high risk of race conditions and high maintenance burden. |
```

Comment:
I see you found the meat of the discussion/PR here! 😉 I appreciate the consideration and emphasis on "unless". Our constraints here are slow, fragile tests vs quality currently. I appreciate that we're choosing quality. The challenge is how to go faster and stable with quality (I want both please!).  I think the answer is in your words of (to avoid) "whole CHT setup". 

I very much like the idea of isolating components and then testing them, and it's cool to have new work showing such an approach already! Testing strategy aside, it builds upon the already-good practice of separation of concerns, and the general pattern recommended by web components. As a related example, I like development of front-end components with things like [Storybook](https://storybook.js.org/) where you build that way, in small parts, and then of course can test that way too. It's a nice side effect that comes with such efforts, which is a funny way to put it because often people start with something like Storybook if only for the joy of building UI components that way.

Back to the testing aspects... if I understand correctly here, we're still talking about running tests with a full web browser (slow), but avoiding all the risks in a true "end to end" (fragile and full of dependencies) test. Execution time goes up somewhat thanks to less initialization/setup when running tests, but we still have to wait on the browser startup and click events. There are still risks of race conditions, although far fewer since we've cut out all the other parts (like the DB).

_⚠ Warning, sketchy suggestion follows:_
There may be a route to getting the speed attribute here too. If we get our separation of concerns right (at least at the web component level) we could detect what changed and only execute the UI component tests for those. On average, that would mean running almost zero of them. That wouldn't work for an e2e (or even integration) test since those involve too many dependencies to be guaranteed of when it's safe to skip running a test. If we're doing a great job on isolated components though it does become safe.

Reply:
I expect the point of this document to be current best practice, rather than desirable future state. Given that, stating a number of total tests that is an order of magnitude smaller than the current suite probably isn't going to help anyone.

It would probably be more useful to talk about when to add a new e2e test, or some sort of talk about coverage in terms of pages covered or similar.

Reply:
It is really hard for me to comment on this topic because I don't have a solution. I love the e2e tests because it is what we do when we used to do manual tests, and I have to be honest, I really like to automate things 😅  But I also agree that it is painful to see how much time takes to run them and how fragile they are.

I loved the way the e2e tests were isolated for the web component, they are faster and as effective as the e2e test tests that we have for the cht-core but without having the complete setup. It would be great if we could have something similar for other sections of the system.

About what Gareth was saying related to when to add new e2e tests (that is something Josh also mentioned in his comment) we have discussed this in our QA syncs, how to have a "rule" or a "guide" that the developers can follow every time that they create a new feature. In a meeting that we had, Josh mentioned that he has a "personal rule", and every time that he changes something in the `webapp` it indicates that a new test (or a new scenario) needs to be added because it will be the only way to actually test it, and I think that it is a really good idea, and easy to follow. 
I don't know that @craig-landry or @garethbowen think about that?

Reply:
> rather than desirable future state

Good point; agreed.

Reply:
It is really hard for me to comment on this topic because I don’t have a solution. I love the e2e tests because it is what we do when we used to do manual tests, and I have to be honest, I really like to automate things :sweat_smile: But I also agree that it is painful to see how much time takes to run them and how fragile they are.
I loved the way the e2e tests were isolated for the web component, they are faster and as effective as the e2e test tests that we have for the cht-core but without having the complete setup. It would be great if we could have something similar for other sections of the system.

About what Gareth was saying related to when to add new e2e tests (that is something Josh also mentioned in his comment) we have discussed this in our QA syncs, how to have a “rule” or a “guide” that the developers can follow every time that they create a new feature. In a meeting that we had, Josh mentioned that he has a “personal rule”, and every time that he changes something in the webapp it indicates that a new test (or a new scenario) needs to be added because it will be the only way to actually test it, and I think that it is a really good idea, and easy to follow.

I don’t know what @craig-landry or @garethbowen think about that?

Reply:
This is the controversial part of the document. 
We mainly test UI features in this layer, and I agree with Josh that if we don´t have a way to test web app features on other layers, it will be challenging to reduce the number of e2e tests. E2e tests give us the most confidence to decide if your feature is working; we need to start changing the current setup if we want to change that fact.
We can also make some effort about duplicate scenarios; we need to clean up the scenarios covered on our e2e test and define main workflows that cover multiple features or something similar to try to reduce the current number of tests. This cleanup may also help when deciding whether to add a whole new test or some extra steps to existing ones.